src/proxy/utils.ts

```diff
@@ -1,106 +1,42 @@
 import { IncomingHttpHeaders, IncomingMessage, ServerResponse } from 'http'
-import { Readable } from 'stream'
-import * as zlib from 'zlib'
 
 import undici from 'undici'
+import { HttpMethod } from 'undici/types/dispatcher'
 
 import { config } from '../config'
 import { logger } from '../logger'
 
-import { IProxyRouteConfig, IProxyRouteMethod } from './types'
+import { IProxyRouteConfig, IStreamOpaque } from './types'
 
-export const buildHostUrl = (): string | null => {
-	const host = config.get('proxy.hosts.relayProxy') as string | null
-
-	if (host) {
-		if (host.startsWith('https://')) {
-			return host
-		}
-		return `https://${host}`
-	}
-
-	return null
-}
-
-export const bufferToJson = (buffer: Buffer): unknown => {
-	try {
-		return JSON.parse(buffer.toString('utf-8'))
-	} catch (err) {
-		logger.error(
-			{ err, buffer: buffer.toString('utf-8') },
-			'Failed to parse json'
-		)
-		return null
+export const buildUrl = (route: IProxyRouteConfig, reqUrl: string): string => {
+	if (route.stripPrefix) {
+		return `${route.targetHost}${reqUrl.substring(route.prefix.length)}`
 	}
-}
-
-const jsonToBuffer = (json: unknown): Buffer => {
-	try {
-		return Buffer.from(JSON.stringify(json))
-	} catch (err) {
-		logger.error({ err, json }, 'Failed to stringify json')
-		return Buffer.from('{ "error": "Failed to process upstream response" }')
-	}
-}
-
-export const compressJson = async (json: unknown): Promise<Buffer> => {
-	return new Promise((resolve, reject) => {
-		zlib.gzip(jsonToBuffer(json), (err, buffer) => {
-			if (err) {
-				logger.error({ err, json }, 'Failed to compress json')
-				reject(err)
-			}
-			resolve(buffer)
-		})
-	})
-}
-
-export const decompressResponse = async (stream: Readable): Promise<Buffer> => {
-	const gunzip = zlib.createGunzip()
-	stream.pipe(gunzip)
-	return new Promise((resolve) => {
-		const buffers: Buffer[] = []
-
-		gunzip
-			.on('data', (chunk: Buffer) => {
-				buffers.push(chunk)
-			})
-			.on('end', () => {
-				resolve(Buffer.concat(buffers))
-			})
-			.on('error', (err: Error) => {
-				logger.error({ err }, 'Failed to decompress response')
-				resolve(
-					Buffer.from('{ "error": "Failed to decompress response" }')
-				)
-			})
-	})
-}
-
-const stripPrefix = (reqUrl: string, prefix?: string): string => {
-	if (prefix && reqUrl.startsWith(prefix)) {
-		return reqUrl.substring(prefix.length)
-	}
-
-	return reqUrl
+	return `${route.targetHost}${reqUrl}`
 }
 
 export const prepareHeaders = (
+	url: string,
 	originalHeaders: IncomingHttpHeaders,
 	remoteAddress?: string
 ): IncomingHttpHeaders => {
+	const webWriterHost = config
+		.get('proxy.hosts.webSdkWriter')
+		.split('https://')[1]
+
+	const host = url.includes('/rec/v2/write')
+		? webWriterHost
+		: // Next line bypasses: [ERR_TLS_CERT_ALTNAME_INVALID]: Hostname/IP does not match certificate's altnames.
+		  ''
+
```

Comment:
I would rather propagate the host header from the "routes" config and set it for all services.